inferno-vc/src/Inferno/VersionControl/Operations.hs

```diff
@@ -1,439 +1,50 @@
-                  -- with the same hash. This way the upstream system (e.g. onping/frontend) can differentiate between
-                  -- source that is still available and no longer available.
-                  -- This does not change the way the script is persisted in the db, it is still stored as 'CloneOf'.
-                  -- See 'CloneOfRemoved' for details.
-                  pure [obj {Inferno.VersionControl.Types.pred = CloneOfRemoved hsh'}, ori]
-                Nothing ->
-                  -- This script no longer exists even in 'removed' directory. The directory might get cleaned up by accident or something.
-                  -- There are two choices we can make,
-                  -- 1. Return a `VCMeta VCObjectHash` with dummy data
-                  -- 2. Ignore this meta.
-                  -- Approach no. 2 is taken here
-                  getHistory history >>= \res -> pure $ obj : res
-            _ -> getHistory history >>= \res -> pure $ obj : res
-    getHistory [] = pure []
+  -- | Fetch all dependencies of an object.
+  fetchVCObjectClosureHashes :: VCObjectHash -> m [VCObjectHash]
 
-    getObj hsh = do
-      VCStorePath storePath <- asks getTyped
-      existsInRoot <- liftIO $ doesFileExist $ storePath </> show hsh
-      if existsInRoot
-        then do
-          obj <- fmap (const hsh) <$> fetchVCObject hsh
-          pure $ Just $ Right obj
-        else do
-          existsInRemoved <- liftIO $ doesFileExist $ storePath </> "removed" </> show hsh
-          if existsInRemoved
-            then do
-              obj <- fmap (const hsh) <$> fetchRemovedVCObject hsh
-              pure $ Just $ Left obj
-            else pure Nothing
+  fetchVCObjectHistory :: (Deserializable m a, Deserializable m g) => VCObjectHash -> m [VCMeta a g VCObjectHash]
 
-getAllHeads :: (VCStoreLogM env m, VCStoreEnvM env m) => m [VCObjectHash]
-getAllHeads = do
-  VCStorePath storePath <- getTyped <$> ask
-  -- We don't need a lock here because this only lists the heads/ directory, it doesn't
-  -- read any file contents (and I assume the `ls` is atomic)
-  headsRaw <- liftIO $ getDirectoryContents $ storePath </> "heads"
-  pure $
-    foldr
-      ( \hd xs ->
-          case (either (const $ Nothing) Just $ Base64.decode $ Char8.pack hd) >>= digestFromByteString of
-            Just hsh -> (VCObjectHash hsh) : xs
-            Nothing -> xs
-      )
-      []
-      (map takeFileName headsRaw)
+  -- | Fetch all objects that are public or that belong to the given set of groups.
+  -- Note this is a potentially long operation so no locks are held while traversing the
+  -- store and checking every object making this operation weakly consistent.
+  -- This means the returned list does not necessarily reflect the state of the store at any
+  -- point in time.
+  fetchFunctionsForGroups :: (Ord g, Deserializable m a, Deserializable m g) => Set.Set g -> m [VCMeta a g VCObjectHash]
```

Comment:
I left it here since the filesystem backend does not guarantee the result being a hard snapshot, although a sqlite backend could with transactions. Better stick to the LCD in the interface imho